crates/provider/src/docker/client.rs

```diff
@@ -0,0 +1,444 @@
+where
+    D: Deserializer<'de>,
+{
+    let str_sequence = String::deserialize(deserializer)?;
+    Ok(str_sequence
+        .split(',')
+        .filter(|item| !item.is_empty())
+        .map(|item| item.to_owned())
+        .collect())
+}
+
+fn deserialize_null_as_default<'de, D, T>(deserializer: D) -> std::result::Result<T, D::Error>
+where
+    T: Default + Deserialize<'de>,
+    D: Deserializer<'de>,
+{
+    let opt = Option::deserialize(deserializer)?;
+    Ok(opt.unwrap_or_default())
+}
+
+impl ContainerRunOptions {
+    pub fn new<S>(image: &str, command: Vec<S>) -> Self
+    where
+        S: Into<String> + std::fmt::Debug + Send + Clone,
+    {
+        ContainerRunOptions {
+            image: image.to_string(),
+            command: command
+                .clone()
+                .into_iter()
+                .map(|s| s.into())
+                .collect::<Vec<_>>(),
+            env: None,
+            volume_mounts: None,
+            name: None,
+            entrypoint: None,
+            rm: false,
+        }
+    }
+
+    pub fn env<S>(mut self, env: Vec<(S, S)>) -> Self
+    where
+        S: Into<String> + std::fmt::Debug + Send + Clone,
+    {
+        self.env = Some(
+            env.into_iter()
+                .map(|(name, value)| (name.into(), value.into()))
+                .collect(),
+        );
+        self
+    }
+
+    pub fn volume_mounts<S>(mut self, volume_mounts: HashMap<S, S>) -> Self
+    where
+        S: Into<String> + std::fmt::Debug + Send + Clone,
+    {
+        self.volume_mounts = Some(
+            volume_mounts
+                .into_iter()
+                .map(|(source, target)| (source.into(), target.into()))
+                .collect(),
+        );
+        self
+    }
+
+    pub fn name<S>(mut self, name: S) -> Self
+    where
+        S: Into<String> + std::fmt::Debug + Send + Clone,
+    {
+        self.name = Some(name.into());
+        self
+    }
+
+    pub fn entrypoint<S>(mut self, entrypoint: S) -> Self
+    where
+        S: Into<String> + std::fmt::Debug + Send + Clone,
+    {
+        self.entrypoint = Some(entrypoint.into());
+        self
+    }
+
+    pub fn rm(mut self) -> Self {
+        self.rm = true;
+        self
+    }
+}
+
+impl DockerClient {
+    pub async fn new() -> Result<Self> {
+        let using_podman = Self::is_using_podman().await?;
+
+        Ok(DockerClient { using_podman })
+    }
+
+    async fn is_using_podman() -> Result<bool> {
+        let result = tokio::process::Command::new("docker")
+            .arg("--version")
+            .output()
+            .await
+            .map_err(|err| anyhow!("Failed to detect container engine: {err}"))?;
+
+        if !result.status.success() {
+            return Err(anyhow!(
+                "Failed to detect container engine: {}",
+                String::from_utf8_lossy(&result.stderr)
+            )
+            .into());
+        }
+
+        Ok(String::from_utf8_lossy(&result.stdout).contains("podman"))
```

Comment:
This will require that the user have at least a symlink `docker -> podman`, why not:
- Allow users to pass the `main cmd` (docker/podman) in `fn new`
- or, detect one or the other. Trying first `docker` and then `podman`

Thx!